### PR TITLE
[jit] dict values(), keys(), and len()

### PR DESCRIPTION
**Pull request by driazati**
Adds some operations for dicts to match Python and tests

### CODE REVIEW

File: torch/csrc/jit/operator.cpp [hunk 1]
```diff
@@ -160,6 +160,15 @@ struct SchemaParser {
       L.next();
       value = DynamicType::get();
       alias_info = parseAliasAnnotation();
+    } else if (L.cur().kind == TK_IDENT && L.cur().text() == "Dict") {
+      L.next();
+      L.expect('[');
```
Comment (zdevito) [inline torch/csrc/jit/operator.cpp]: discussion: change to '(' to match future. Try to see if we can normalize syntax.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 2]
```diff
@@ -1205,6 +1199,54 @@ Operation listSetItem<Shared<BoolList>, bool>(const Node* node) {
   };
 }
 
+Operation dictLen(const Node* node) {
```
Comment (zdevito) [inline torch/csrc/jit/register_prim_ops.cpp]: see if you can just have the direct function rather than the one that takes a Node

**Review by zdevito — APPROVED** [review]
Looks good!

**Comment by ezyang** [comment]
This PR broke ASAN and is being reverted.

File: aten/src/ATen/core/jit_type.h [hunk 3]
```diff
@@ -560,7 +560,11 @@ struct DictType : public Type {
   }
 
   DEFINE_IS_SUBCLASS(DictType);
+
   bool isSubtypeOf(const TypePtr rhs) const override {
+    if (rhs->kind() != TypeKind::DictType) {
```
Comment (zdevito) [inline aten/src/ATen/core/jit_type.h]: I am unclear on why this would fix asan, can you clarify?

**Comment by sidazhang** [comment]
Does this allow us to create dictionaries inside torchscript too?

File: aten/src/ATen/core/jit_type.h (hunk 3 above)
Comment (driazati) [reply aten/src/ATen/core/jit_type.h]: It won't, #16534 has the correct fix

**Comment by driazati** [comment]
@sidazhang literal dicts are supported (type is inferred from the elements if they're not empty, `torch.jit.annotate` must be used if you want to declare an empty dict other than the default of `Dict[str, Tensor]`):

https://github.com/pytorch/pytorch/blob/8f3bf9b8b3ee0a29d0f4ed0fa288a09ce0f5db4e/test/test_jit.py#L9808-L9811

**Comment by sidazhang** [comment]
Got it. this is not mutable right? I cannot update the dict once it is created

**Comment by driazati** [comment]
At the moment yes, but mutability is coming soon

**Pull request closed without merging**